### PR TITLE
Initial skeleton changes + model and dataset download flow

### PR DESCRIPTION
**Pull request by watiss**

### CODE REVIEW

File: docs/conf.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 HERE = Path(__file__).parent
 sys.path[:0] = [str(HERE.parent), str(HERE / "extensions")]
 
-import mypackage  # noqa
+import scvimadz  # noqa
```
Comment (jjhong922) [inline docs/conf.py:15]: curious, what does madz stand for?

File: scvimadz/reference/_tabula_sapiens.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+from typing import Type
+
+import pandas as pd
+from reference.base import BaseReference
+from scvi.model.base import BaseModelClass
+from storage import ZenodoStorage
+
+
+class TabulaSapiensReference(BaseReference):
+    def __init__(self):
```
Comment (jjhong922) [inline scvimadz/reference/_tabula_sapiens.py:10]: This should probably take a zenodo URL. If we end up supporting different storage options, we could allow for different storage identifier kwargs and assert that exactly one is provided.

File: scvimadz/reference/_tabula_sapiens.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        self._datasets = {}  # id to dataset
+        self._init_models()
+        self._init_datasets()
+
+    def _init_models(self) -> None:
+        # TODO update `if` condition depdending on how we will store the data on Zenodo
+        self._model_keys = [
+            key
+            for key in self._store.list_model_keys()
+            if key.contains("tabula_sapiens")
```
Comment (jjhong922) [inline scvimadz/reference/_tabula_sapiens.py:22]: Is the idea that the repo could potentially contain other things? I was thinking that the repo would be expected to be tabula sapiens specific already

File: scvimadz/reference/_tabula_sapiens.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        ]
+        for model_key in self._model_keys:
+            self.load_model(model_key)
```
Comment (jjhong922) [inline scvimadz/reference/_tabula_sapiens.py:25]: This will be really slow if we try to load them all during init, should wait until the user calls for a model specifically to load it.

File: scvimadz/storage/_zenodo.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+from typing import List
+
+from storage.base import BaseStorage
+
+
+class ZenodoStorage(BaseStorage):
+    def __init__(self, url: str):
+        self._url = url
+
+    def list_model_keys(self) -> List[str]:
```
Comment (jjhong922) [inline scvimadz/storage/_zenodo.py:10]: I would avoid having the idea of a model and dataset even present in this class, and keep it to files and keys. Right now since theres the idea of a model and a dataset in both the reference and storage class, it's unclear what logic to put in each since there is no clear abstraction layer which could lead to it being messy later in the implementation.

**Review by jjhong922 — COMMENTED** [review]
Some initial comments, good work so far!

File: docs/conf.py (hunk 1 above)
Comment (watiss) [reply docs/conf.py:15]: Model And Data Zoo :D

File: scvimadz/reference/_tabula_sapiens.py (hunk 2 above)
Comment (watiss) [reply scvimadz/reference/_tabula_sapiens.py:10]: I aimed to conceal the notion of storage behind the reference so users dont have to worry about providing that when instantiating a reference. But that also means that this TSReference is coupled with a Zenodo backend the way it's written. If we want a TSRefernece with a different backend, we can create a base TSReference and derive two child classes off of it each of which has their own backend.
We can also expose Storage at Reference creation time as you suggest. It isn't the design I went for but am open to exploring it. If we go with that, I'd prefer providing a Storage kwarg rather than a raw url string though.
Curious to know you guys' thoughts

File: scvimadz/reference/_tabula_sapiens.py (hunk 3 above)
Comment (watiss) [reply scvimadz/reference/_tabula_sapiens.py:22]: yes I was thinking that the repo could contain other pre-trained models, e.g. models pre-trained on other datasets.

File: scvimadz/reference/_tabula_sapiens.py (hunk 4 above)
Comment (watiss) [reply scvimadz/reference/_tabula_sapiens.py:25]: Fair enough. I considered lazy loading but thought I might as well do it all at once. But if model loading is slow as you mention, that's not a good idea.

File: scvimadz/storage/_zenodo.py (hunk 5 above)
Comment (watiss) [reply scvimadz/storage/_zenodo.py:10]: Yeah that's a good call. I'll update.

File: scvimadz/reference/_tabula_sapiens.py (hunk 2 above)
Comment (jjhong922) [reply scvimadz/reference/_tabula_sapiens.py:10]: Ah ok makes sense, didn't realize the plan was to hardcode the URL for this reference. That is fine for now I think. More general reference classes can offer a solution to this.

File: scvimadz/reference/base/_base_reference.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+import rich_dataframe
+from anndata import AnnData
+from scvi.model.base import BaseModelClass
+from storage.base import BaseStorage
+
+
+class BaseReference(ABC):
+    def __init__(self) -> None:
+        super().__init__()
+        self._reference_prefix = self.reference_name + "_"  # e.g. tabula_sapiens_
```
[the remark this answers is not in the document]
Comment (jjhong922) [reply scvimadz/reference/base/_base_reference.py:17]: also this is a little weird to me bc self.reference_name is an instance method and this is in the instance init method which means it could potentially fail if reference_name is defined as a function of an instance property. Alternatively you can just have a concrete instance method for this:

`def reference_prefix(self): return f"{self.reference_name}_"`

File: scvimadz/reference/base/_base_reference.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        pass
+
+    @property
+    @abstractmethod
+    def data_store(self) -> Type[BaseStorage]:
+        """ The backend store for datasets """
+        pass
+
+    def _get_object_keys(self, obj_type: str) -> List[str]:
+        assert obj_type in ["model", "dataset"]
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:38]: check and raise a ValueError here instead of an assert. Also, define {"model", "dataset"} as a constant set at the top of the file named `_OBJ_TYPES`.

File: scvimadz/reference/base/_base_reference.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        store = self.model_store if obj_type == "model" else self.data_store
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:39]: you can also define these constants individually since you use separately like here

File: scvimadz/reference/base/_base_reference.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        return keys
+
+    def _list_objects(
+        self, obj_type: str, metadata_fn: str, pretty_print: bool = False
+    ) -> pd.DataFrame:
+        keys = self._get_object_keys(obj_type)
+        metadata_file = self.model_store.download_file(metadata_fn)
+        df = pd.read_csv(metadata_file, index_col="key").loc[keys]
+        if pretty_print:
+            rich_dataframe.prettify(
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:52]: does this print to console or do you need to return it?

File: scvimadz/reference/base/_base_reference.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                df,
+                row_limit=len(df),
+                col_limit=len(df.columns),
+                clear_console=False,
+            )
+        else:
+            return df
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:59]: reason not to return df in both cases?

File: scvimadz/reference/base/_base_reference.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+
+    def list_models(self, pretty_print: bool = False) -> pd.DataFrame:
+        """ Lists all available models associated with this reference """
+        return self._list_objects("model", "models_metadata.csv", pretty_print)
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:63]: save these magic strings as constants since they are fixed file names

File: scvimadz/reference/base/_base_reference.py [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+
+    def list_datasets(self, pretty_print: bool = False) -> pd.DataFrame:
+        """ Lists all available datasets associated with this reference """
+        return self._list_objects("dataset", "datasets_metadata.csv", pretty_print)
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:67]: same here

File: scvimadz/reference/base/_base_reference.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        # get the cell that contains the cls_name for this model, it will be like: scvi.model.TOTALVI
+        model_cls_name = models.loc[model_id, "cls_name"]
+        cls = model_cls_name.split(".")[-1]
+        module = ".".join(model_cls_name.split(".")[:-1])
+        if adata is None:
+            model_adata = models.loc[model_id, "train_dataset"]
+            adata = self.load_dataset(model_adata)
+        model_cls = getattr(importlib.import_module(module), cls)
+        model_file = self.model_store.download_file(model_id)
+        model_cls.load(os.path.dir_name(model_file), adata=adata, use_gpu=use_gpu)
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:103]: do you need to return this?

File: scvimadz/reference/base/_base_reference.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+import anndata
+import pandas as pd
+import rich_dataframe
+from anndata import AnnData
+from scvi.model.base import BaseModelClass
+from storage.base import BaseStorage
+
+
+class BaseReference(ABC):
+    def __init__(self) -> None:
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:15]: for model_store and data_store I would add them as arguments here instead of creating abstract properties for them. Then call super().__init__() w/ them in each child class init

File: scvimadz/reference/_tabula_sapiens.py [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+class TabulaSapiensReference(BaseReference):
+    def __init__(self, data_dir: str):
+        # TODO replace 961966 (our sandbox record id) with our real repo id's
+        self._model_store = ZenodoStorage("962638", data_dir)
+        self._data_store = ZenodoStorage("962630", data_dir)
+        super().__init__()
+
+    @property
+    def reference_name(self) -> str:
+        return "tabula_sapiens"
```
Comment (jjhong922) [inline scvimadz/reference/_tabula_sapiens.py:18]: nit: you can just do `reference_name = "tabula_sapiens"` at the top of the class instead

File: scvimadz/storage/_zenodo.py [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    def __init__(self, record_id: str, data_dir: str, sandbox: bool = False):
+        self._record_id = record_id
+        self._data_dir = data_dir
+        if not os.path.isdir(data_dir):
+            raise ValueError(f"Error: {data_dir} is not a valid directory")
+        # zenodo urls
+        self._zenodo_base_url = (
+            "https://zenodo.org/" if not sandbox else "https://sandbox.zenodo.org/"
+        )
+        self._zenodo_record_url = self._zenodo_base_url + "record/{}/files/{}"
```
Comment (jjhong922) [inline scvimadz/storage/_zenodo.py:19]: nit: use f-strings here too

Comment (jjhong922) [reply scvimadz/storage/_zenodo.py:19]: would name this `_zenodo_record_url_template` to explicit

**Review by jjhong922 — COMMENTED** [review]
some nits and suggestions. feel free to ping me if you have questions on anything!

File: scvimadz/reference/base/_base_reference.py (hunk 14 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:15]: That's actually what I had in a prev iteration (might not have pushed it to GitHub). I ended up choosing this impl with abstract props instead in order to enforce subclassing BaseReference. That seems more in line with the pattern we have elsewhere...

File: scvimadz/reference/_tabula_sapiens.py (hunk 15 above)
Comment (watiss) [reply scvimadz/reference/_tabula_sapiens.py:18]: hmm, but I wanted to have abtract props (see other related comment). Can I do what you suggest while keeping that functionality?

File: scvimadz/reference/base/_base_reference.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:17]: Hmm I am not sure I follow. Why would this fail if `reference_name` is defined as a function of an instance property?

File: scvimadz/reference/base/_base_reference.py (hunk 10 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:59]: Oh good catch. Probably a leftover from a previous test I was doing.

File: scvimadz/reference/base/_base_reference.py (hunk 9 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:52]: yep it prints

File: scvimadz/reference/base/_base_reference.py (hunk 13 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:103]: I indeed do :D Good catch. (still testing this method)

File: scvimadz/reference/base/_base_reference.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jjhong922) [reply scvimadz/reference/base/_base_reference.py:17]: reference_name is an instance method, which means the developer is free to make it a function of anything relative to `self`. So a natural thing to do would be adding a reference_name argument to the init statement and then assigning that to `self._reference_name` and having the method return that. However, if you do that, this prefix thing will error because prior to assigning that variable, `self.reference_name` will be None.

[the remark this answers is not in the document]
Comment (jjhong922) [reply scvimadz/reference/base/_base_reference.py:17]:
```
class Example(BaseReference):
  def __init__(self, ref_name):
    super().__init__()
    self._reference_name = ref_name

  @property
  def reference_name(self):
    return self._reference_name

e = Example() # errors in __init__ bc you cannot do None + "_"
```

File: scvimadz/reference/base/_base_reference.py (hunk 14 above)
Comment (jjhong922) [reply scvimadz/reference/base/_base_reference.py:15]: We can still have the properties but just put them into the BaseReference class. By adding them as args to the BaseReference init method it'll still enforce child classes to pass a model and data store in. Since it's an abstract class, subclassing is always required.

Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:21]: Talked offline. We would still need the abstract methods to enforce subclassing (merely inheriting from ABC does not suffice).

File: scvimadz/storage/_zenodo.py (hunk 16 above)
Comment (watiss) [reply scvimadz/storage/_zenodo.py]: I switched to using f-strings for the vars below this line but for `_zenodo_record_url` I could not use f-strings since it would complain about the other two missing placeholders (please let me know if there is a better way to do it, I was just doing: `f"{self._zenodo_base_url}record/{}/files/{}"`).

File: scvimadz/reference/base/_base_reference.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py]: Talked offline. I was counting on `super().__init__()` being called at the end of the child init method in order to avoid these issues. But it looks like the common Python pattern is to call super init at the beginning, which could break things. I'll try the suggestion above.

File: scvimadz/storage/_zenodo.py (hunk 16 above)
Comment (jjhong922) [reply scvimadz/storage/_zenodo.py]: oh yea for templates f strings require double brackets like this:
f"{self._zenodo_base_url}record/{{}}/files/{{}}"
then it resolves to single brackets

File: scvimadz/reference/base/_base_reference.py [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    @property
+    @abstractmethod
+    def model_store(self) -> Type[BaseStorage]:
+        """ The backend store for models """
+        pass
+
+    @property
+    @abstractmethod
+    def data_store(self) -> Type[BaseStorage]:
+        """ The backend store for datasets """
```
Comment (adamgayoso) [inline scvimadz/reference/base/_base_reference.py:40]: we should add codacy to this repo if it's not there already. This shouldn't have spaces around and should have a period

File: tests/test_references.py [hunk 18]
```diff
@@ -0,0 +1,7 @@
+from scvimadz.reference import TabulaSapiens
+
+
+def test_ts_reference():
+    model = TabulaSapiens()
```
Comment (adamgayoso) [inline tests/test_references.py:5]: is there a way we can test this thing with zenodo? Probably not, right? We could write a `LocalStorage` class in the next pr to get some local testing ?

**Review by adamgayoso — CHANGES_REQUESTED** [review]
are we still on track to use this API to put models onto storage?

File: tests/test_references.py (hunk 18 above)
Comment (watiss) [reply tests/test_references.py:5]: I've added some network tests. I think a `LocalStorage` might be a good idea as well. I will look into it, if not in this PR in the next where I will implement `save`.

**Comment by watiss** [comment]
> are we still on track to use this API to put models onto storage?

Yes

File: scvimadz/_utils.py [hunk 19]
```diff
@@ -0,0 +1,6 @@
+def dirname(path: str) -> str:
```
[the remark this answers is not in the document]
Comment (watiss) [reply scvimadz/_utils.py:1]: Last time I tried that, it did not work for me.. but let me try again.

[the remark this answers is not in the document]
Comment (watiss) [reply scvimadz/_utils.py]: It works! I deleted _utils.py

File: pyproject.toml [hunk 20]
```diff
@@ -78,3 +79,8 @@ requires = [
   "poetry>=1.0",
   "setuptools", # keep it here or "pip install -e" would fail
 ]
+
+[tool.pytest.ini_options]
+markers = [
+    "network-tests: tests that involve network operations",
+]
```
Comment (jjhong922) [inline pyproject.toml:86]: nit: newline

File: scvimadz/reference/base/_base_reference.py [hunk 21]
```diff
@@ -0,0 +1,137 @@
...
+    @abstractmethod
+    def data_store(self) -> Type[BaseStorage]:
+        """The backend store for datasets."""
+        pass
+
+    def _get_reference_prefix(self) -> str:
+        return f"{self.reference_name}_"  # e.g. tabula_sapiens_
+
+    def _get_store_for_object(self, obj_type: str) -> Type[BaseStorage]:
+        if obj_type not in [_OBJ_TYPE_MODEL, _OBJ_TYPE_DATASET]:
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:44]: If this string is only limited to two strings, you should explore using enums: https://docs.python.org/3/library/enum.html.

File: scvimadz/reference/base/_base_reference.py [hunk 22]
```diff
@@ -0,0 +1,137 @@
...
+        Parameters
+        ----------
+        dataset_id
+            id of the dataset to load
+
+        Returns
+        -------
+        An instance of :class:`~anndata.AnnData` associated with the given dataset id.
+        """
+        data_file = self.data_store.download_file(dataset_id)
```
Comment (jjhong922) [inline scvimadz/reference/base/_base_reference.py:136]: Is the return here a filepath? If so it would make more sense to call this data_filepath

File: scvimadz/storage/_zenodo.py [hunk 23]
```diff
@@ -0,0 +1,57 @@
...
+        """Returns all keys in this storage."""
+        response = requests.get(self._zenodo_api_records_url + self._record_id)
+        # for the status codes Zenodo uses, see https://developers.zenodo.org/#responses
+        response.raise_for_status()
+        # if the call above didn't throw the response was "ok" (code < 400)
+        response_json = response.json()
+        keys = [elem["key"] for elem in response_json["files"]]
+        return keys
+
+    def download_file(self, key: str) -> str:
```
Comment (jjhong922) [inline scvimadz/storage/_zenodo.py:35]: What is key vs id here?

File: scvimadz/storage/base/_base_storage.py [hunk 24]
```diff
@@ -0,0 +1,25 @@
+from abc import ABC, abstractmethod
+from typing import List
+
+
+class BaseStorage(ABC):
```
Comment (jjhong922) [inline scvimadz/storage/base/_base_storage.py:5]: Can you add a docstring describing the role of a Storage class and what keys are?

File: tests/mock/models/models_metadata.csv [hunk 25]
```diff
@@ -0,0 +1,3 @@
+key,cls_name,hyperparameter_alpha,hyperparameter_beta,batch_corrected,train_dataset
+hca_model_80262d08-4a30-4071-a3c6-96274182646d.zip,scvi.model.SCVI,0.5,0.9,Yes,hca_dataset_dcfaad7a-70a4-4669-87d2-7bb241673097.h5ad
```
Comment (jjhong922) [inline tests/mock/models/models_metadata.csv:2]: do we expect to have .zip in the key?

**Review by jjhong922 — COMMENTED** [review]
Few minor comments, looks good to move onto creating an API for uploading a model/dataset

File: tests/mock/models/models_metadata.csv (hunk 25 above)
Comment (watiss) [reply tests/mock/models/models_metadata.csv:2]: Yes that's the scheme I've been using so far

File: scvimadz/storage/_zenodo.py (hunk 23 above)
Comment (watiss) [reply scvimadz/storage/_zenodo.py:35]: They're the same thing, I fixed the comment

File: scvimadz/reference/base/_base_reference.py (hunk 21 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:44]: Do you mean something like:
```
class ObjType(Enum):
...     MODEL = "model"
...     DATASET = "dataset"
```
?
What is the benefit of that?

File: scvimadz/reference/base/_base_reference.py (hunk 22 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py:136]: yes filepath

File: scvimadz/reference/base/_base_reference.py (hunk 21 above)
Comment (jjhong922) [reply scvimadz/reference/base/_base_reference.py:44]: It allows you to type the obj_type argument insteading of having it as an arbitrary string type. It is more clear from the signature that it can only be two values rather than any string

File: scvimadz/reference/base/_base_reference.py [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+        if pretty_print:
+            rich_dataframe.prettify(
+                df,
+                row_limit=len(df),
+                col_limit=len(df.columns),
+                clear_console=False,
+            )
+        return df
+
+    def list_models(self, pretty_print: bool = False) -> pd.DataFrame:
```
Comment (adamgayoso) [inline scvimadz/reference/base/_base_reference.py:83]:
```suggestion
    def get_models_df(self, pretty_print: bool = False) -> pd.DataFrame:
```
list to me seems like it would return a list

File: scvimadz/reference/base/_base_reference.py [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+        An instance of :class:`~scvi.model.base.BaseModelClass` associated with the given model id.
+        """
+        models = self.list_models()
+        # get the cell that contains the cls_name for this model, it will be like: scvi.model.TOTALVI
+        model_cls_name = models.loc[model_id, "cls_name"]
+        cls = model_cls_name.split(".")[-1]
+        module = ".".join(model_cls_name.split(".")[:-1])
+        if adata is None:
+            model_adata = models.loc[model_id, "train_dataset"]
+            adata = self.load_dataset(model_adata)
```
Comment (adamgayoso) [inline scvimadz/reference/base/_base_reference.py:126]: would be nice to have a comment here

File: tests/mock/_mock_storage.py [hunk 28]
```diff
@@ -0,0 +1,44 @@
+import os
+import shutil
+from pathlib import Path
+from typing import List
+
+from scvimadz.storage.base import BaseStorage
+
+
+class MockStorage(BaseStorage):
+    def __init__(self, data_type: str, save_path: str) -> None:
```
Comment (adamgayoso) [inline tests/mock/_mock_storage.py:10]: you can use the Literal type we define in scvi for data_type

File: tests/mock/models/models_metadata.csv [hunk 29]
```diff
@@ -0,0 +1,3 @@
...
+tabula_sapiens_model_bfeb6529-8db7-46ce-99f6-237748798ae4.zip,scvi.model.TOTALVI,0.3,0.8,No,tabula_sapiens_dataset_b4c16899-e458-4cf5-947e-bed59031f156.h5ad
```
Comment (adamgayoso) [inline tests/mock/models/models_metadata.csv:3]: I don't think we should add these binary files to the repo.... It will bloat the cloning. I was thinking the MockStorage testing would involve the save model functionality, so it's all sort of done on the fly during testing. Or, the data we use here has to be trivially small. Will need to edit the git history to remove these files

File: tests/mock/models/models_metadata.csv (hunk 25 above)
Comment (adamgayoso) [reply tests/mock/models/models_metadata.csv:2]: can you explain the concern? @jjhong922

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: tests/mock/models/models_metadata.csv (hunk 25 above)
Comment (jjhong922) [reply tests/mock/models/models_metadata.csv:2]: Just thinking that the ID should be separated from the idea of a filepath. For example if we have unzipped or differently compressed versions of a model, we should be able to append a file type ending to the ID rather than stripping the .zip everytime.

File: tests/mock/_mock_storage.py (hunk 28 above)
Comment (watiss) [reply tests/mock/_mock_storage.py:10]: I think the use of Enum class, as Justin suggested in another comment, obviates the need for this

File: tests/mock/models/models_metadata.csv (hunk 29 above)
Comment (watiss) [reply tests/mock/models/models_metadata.csv:3]: I went with the second option to avoid network traffic. Network roundtrips take a while anyway for large files so we'd anyway have to cut down on the size. We might as well avoid the traffic altogether and keep the small files locally. I cut both datasets down to 50-100 cells and same number of vars (of course they're meaningless now but it's ok for our purposes).
I'll add in a LocalStorage class, peer to ZenodoStorage in the next pr when I add in save functionality

File: scvimadz/reference/base/_base_reference.py (hunk 26 above)
Comment (watiss) [reply scvimadz/reference/base/_base_reference.py]: updated

**Comment by watiss** [comment]
@adamgayoso Could you take another pass?

**Review by adamgayoso — APPROVED** [review]
